commons/all/src/main/java/org/infinispan/commons/util/ProgressTracker.java

```diff
@@ -0,0 +1,142 @@
+package org.infinispan.commons.util;
+
+import java.time.Instant;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.commons.time.TimeService;
+
+import net.jcip.annotations.GuardedBy;
+import net.jcip.annotations.ThreadSafe;
+
+@ThreadSafe
+public final class ProgressTracker {
+
+   private static final Log log = LogFactory.getLog(ProgressTracker.class);
+
+   private final String name;
+   private final ScheduledExecutorService executor;
+   private final TimeService timeService;
+   private final long delay;
+   private final TimeUnit unit;
+
+   @GuardedBy("state")
+   private final State state;
+
+   @GuardedBy("state")
+   private ScheduledFuture<?> progression;
+
+   public ProgressTracker(String name, ScheduledExecutorService executor, TimeService timeService, long delay, TimeUnit unit) {
+      this.name = name;
+      this.executor = executor;
+      this.timeService = timeService;
+      this.state = new State();
+      this.delay = delay;
+      this.unit = unit;
+   }
+
+   public void addTasks(long value) {
+      synchronized (state) {
```

Comment:
Shouldn't we avoid synchronized in lieu of other kinds of locks to avoid pinning  (at least until we can base on top of JDK 24+)

Reply:
Good point. I'll update to be on the safe side.